### PR TITLE
Automatically detect and generate Couch reindex migrations

### PR DESCRIPTION
**Pull request by millerdev**
If a PR  changes a Couch design document but does not update the `couchviews.lock` file, the `test-django-migrations.sh` script will fail because `makemigrations` will want to create new a "reindex" migration in the `preindex` app.

:tropical_fish: Review by commit.

### Safety story

Adds automation to ensure a migration is added if a Couch view is changed, which will allow us to safely remove the auto-reindex-on-every-deploy.

### Automated test coverage

The new migration auto-detection patcher has a test to ensure that changes in Django do not disable app-contributed migration detection. Other things added in this PR are exercised by `test-django-migrations.sh`.

### Migrations

The migration in this PR is effectively a no-op. It performs a Couch reindex to ensure that that process is operational from end-to-end.

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/preindex/django_migrations.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    if not os.path.exists(lock_path):
+        return True
+    with open(lock_path, "r") as file:
+        return list(file) != lines
+
+
+def write_designs_lock_file(lines, lock_path=LOCK_PATH):
+    """Write a new .lock file."""
+    with open(lock_path, "w") as file:
+        file.write("".join(lines))
```
Comment (joxl) [inline corehq/preindex/django_migrations.py:88]: nit
```suggestion
        file.writelines(lines)
```

File: corehq/preindex/django_migrations.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            if (sha, viewpath) not in seen:
+                yield f"{sha} {viewpath}\n"
+                seen.add((sha, viewpath))
+
+
+def designs_did_change(lines, lock_path=LOCK_PATH):
+    if not os.path.exists(lock_path):
+        return True
+    with open(lock_path, "r") as file:
+        return list(file) != lines
```
Comment (joxl) [inline corehq/preindex/django_migrations.py:82]: Is the order of `get_preindex_designs()` deterministic? In other words: if the lines only change in _order_ (if that's even possible), does that justify a "the designs have changed" condition?

**Review by joxl — APPROVED** [review]

File: corehq/preindex/django_migrations.py (hunk 1 above)
Comment (millerdev) [reply corehq/preindex/django_migrations.py:88]: lol, that copied from here

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/hqscripts/management/commands/makemigrations.py#L108

Comment (joxl) [reply corehq/preindex/django_migrations.py:88]: :laughing:

File: corehq/preindex/django_migrations.py (hunk 2 above)
Comment (millerdev) [reply corehq/preindex/django_migrations.py:83]: Good catch!

> Is the order of `get_preindex_designs()` deterministic?

Yes. "Designs" are sorted by design path:

https://github.com/dimagi/commcare-hq/blob/ddc897f6256151828b63968bd9bb246e2708e4cf/corehq/preindex/django_migrations.py#L65

However, Couch design doc manifests apparently were apparently not sorted. Fixed in https://github.com/dimagi/commcare-hq/pull/32544/commits/6d503ca991fd619653212f6053d0a736402d035b

**Review by snopoke — COMMENTED** [review]
I think it would be good to test this on staging prior to merging.

**Comment by millerdev** [comment]
Rebased to remove unnecessary patch in `HqdbContext`. Thanks for the test fix @joxl

**Comment by millerdev** [comment]
Deployed to staging, and confirmed that the reindex is run at the end of the `migrate` command as expected. Ready for review.

**Review by joxl — APPROVED** [review]
Offlined with @millerdev regarding my curiosity about the decision to make `should_reindex` a module global (this has a smell), which was my only question here. Looks good.

**Pull request merged**